damnit/migrations.py

```diff
@@ -112,97 +117,125 @@ def migrate_dataarrays(db, db_dir, dry_run):
+
+    # Get all column and variable names
+    column_names = [rec[0] for rec in
+                    db.conn.execute("SELECT name FROM PRAGMA_TABLE_INFO('runs')").fetchall()]
+    variable_names = [name for name in column_names
+                      if name not in ["proposal", "runnr", "start_time", "added_at", "comment"]]
+    #
+    # And then read all run data. This is what we'll need to copy into the new
+    # `run_variables` table.
+    runs = [rec for rec in db.conn.execute("SELECT * FROM runs").fetchall()]
+
+    print(f"Found {len(runs)} runs, with these variables:")
+    print(variable_names)
+    print()
+
+    # Scan HDF5 files to get timestamps (from mtime) & max diff for 1D arrays
+    timestamps = {}  # keys (proposal, run)
+    max_diffs = {}  # keys (proposal, run, variable)
+    for record in runs:
+        proposal = record["proposal"]
+        run_no = record["runnr"]
+        h5_path = db_dir / f"extracted_data/p{proposal}_r{run_no}.h5"
+
+        if not h5_path.exists():
+            print(f"Skipping variables for run {run_no} because {h5_path} does not exist")
+            continue
+
+        timestamps[(proposal, run_no)] = h5_path.stat().st_mtime
+
+        with h5py.File(h5_path, "a") as f:
+            for name in variable_names:
+                if name not in f:
+                    continue
 
-    # Begin a transaction, anything in here should be rolled back if there's an exception
-    with db.conn:
-        # Get all column and variable names
-        column_names = [rec[0] for rec in
-                        db.conn.execute("SELECT name FROM PRAGMA_TABLE_INFO('runs')").fetchall()]
-        variable_names = [name for name in column_names
-                          if name not in ["proposal", "runnr", "start_time", "added_at", "comment"]]
-
-        # And then read all run data. This is what we'll need to copy into the new
-        # `run_variables` table.
-        runs = [rec for rec in db.conn.execute("SELECT * FROM runs").fetchall()]
-
-        print(f"Found {len(runs)} runs, with these variables:")
-        print(variable_names)
-        print()
-
-        # Now delete the runs table
-        if not dry_run:
-            db.conn.execute("DROP TABLE runs")
-        else:
-            print("Would delete the `runs` table")
-
-        # Execute the v1 schema to create all the new tables
-        if not dry_run:
-            db.conn.executescript(V1_SCHEMA)
-        else:
-            print("Would execute the v1 schema")
-
-        # And re-add the data from the old `runs` table
-        if not dry_run:
-            for record in runs:
-                run = dict(zip(column_names, record))
-                proposal = run["proposal"]
-                run_no = run["runnr"]
-                h5_path = db_dir / f"extracted_data/p{proposal}_r{run_no}.h5"
-
-                # Add the run info to the `run_info` table
-                db.conn.execute("""
+                if (ds := main_dataset(f[name])) is None:
```

Comment:
Nope, this is finding the equivalent of the `data` dataset within a NetCDF group made from an xarray DataArray, i.e. the one that's not coordinate labels.

Reply:
I don't get it, how could a `DataArray` without an non-coordinate array even exist? And won't this skip regular ndarrays?

Reply:
Sorry, I didn't explain that very clearly. Let me start again.

This is part of the code to find the `max_diff` for a given variable, which we store in the DB to efficiently highlight where values are changing during a run. We want to compute that for variables which returned a 1D array, whether that's an xarray or a numpy object.

Variables are stored as an HDF5 group containing at least 1 dataset. If it's a NetCDF group, we have an extra dataset for each labelled dimension:

```text
xarray_test2.nc (1 attributes)
└tkgroup
  ├time	[int64: 4] (6 attributes)
  ├space	[UTF-8 string: 3] (4 attributes)
  └mycoolarray	[float64: 4 × 3] (4 attributes)
```

The `main_dataset()` function looks at a group like this and gives us back the main dataset, the one that's not coordinate labels - i.e. `mycoolarray` in the example above. It will also do the right thing with a group made from a Numpy variable.

The `is None` check on this line catches groups with 0 or >1 main variables. Neither should occur, but it seemed reasonable to fail gracefully. v1 allows saving an xarray Dataset, which means you can have >1 main variable in the future.

The code here is assuming that labelled arrays are in NetCDF format, because we call that conversion at the start of `migrate_v0_to_v1`. When you use `--dry-run` that won't actually be the case, so it won't calculate max_diff for labelled 1D arrays. I figured that was probably good enough; dry-run mode is not easy to do really well.

Reply:
Ah ok, thanks that makes sense. Could you add a comment for that? It wasn't clear to me that this will work for regular ndarrays too.

Reply:
Sure, I actually made it a docstring and a warning message if we don't find a main dataset (since we should always be able to). See what you think. :slightly_smiling_face: